### ISSUE TITLE
Azure AD Login - ERR: TokenError: AADSTS50158: External security challenge not satisfied.

### ISSUE THREAD

**Issue by brianhillnovus** [Ylianst/MeshCentral#5626]
**Describe the bug**
Attempting to log in with Azure AD SSO configured results in an error.

**To Reproduce**
1. Configure Azure AD SSO using steps detailed here: https://ylianst.github.io/MeshCentral/meshcentral/#microsoft-azure-active-directory
2. Click on the SSO login button
3. Enter in Azure AD credentials to authenticate
4. White page with "Internal Server Error" at the top.

**Expected behavior**
Login proceeds, user is created if required.

**Server Software (please complete the following information):**
 - OS: Ubuntu 22.04.3 LTS
 - Virtualization: Docker
 - Network: DMZ/WAN
 - Version: 1.1.16
 - Node: Whatever is bundled in the docker image

**Client Device (please complete the following information):**
 - Device: Laptop
 - OS: Windows 11 Pro
 - Network: LAN
 - Browser: Edge
 - MeshCentralRouter Version: N/A

**Remote Device (please complete the following information):**
 - Device: N/A
 - OS: N/A
 - Network: N/A
 - Current Core Version (if known): N/A

**Additional context**
We have configured our Tenant to use Duo MFA instead of the Microsoft MFA solution. Typically when a user logs in to other applications that are using Azure AD SSO, after providing their credentials they're taken to a Duo MFA page and prompted to approve the login using their Duo app on their phone. This isn't even being presented when trying to log in to Meshcentral.

**Docker Console Output**
```
meshcentral  | MeshCentral HTTP redirection server running on port 80.
meshcentral  | MeshCentral v1.1.16, Hybrid (LAN + WAN) mode, Production mode.
meshcentral  | MeshCentral Intel(R) AMT server running on meshcentral.mydomain.com:4433.
meshcentral  | MeshCentral HTTPS server running on meshcentral.mydomain.com:443.
meshcentral  | ERR: TokenError: AADSTS50158: External security challenge not satisfied. User will be redirected to another page or authentication provider to satisfy additional authentication challenges. Trace ID: bc6a457b-5365-4cc8-b6e4-420c2e324100 Correlation ID: 8487efe4-1c92-4d7c-87c2-45be53d1b41a Timestamp: 2023-12-12 21:13:12Z
meshcentral  |     at OAuth2Strategy.parseErrorResponse (/opt/meshcentral/meshcentral/node_modules/passport-oauth2/lib/strategy.js:373:12)
meshcentral  |     at OAuth2Strategy._createOAuthError (/opt/meshcentral/meshcentral/node_modules/passport-oauth2/lib/strategy.js:420:16)
meshcentral  |     at /opt/meshcentral/meshcentral/node_modules/passport-oauth2/lib/strategy.js:177:45
meshcentral  |     at /opt/meshcentral/meshcentral/node_modules/oauth/lib/oauth2.js:191:18
meshcentral  |     at passBackControl (/opt/meshcentral/meshcentral/node_modules/oauth/lib/oauth2.js:132:9)
meshcentral  |     at IncomingMessage.<anonymous> (/opt/meshcentral/meshcentral/node_modules/oauth/lib/oauth2.js:157:7)
meshcentral  |     at IncomingMessage.emit (node:events:529:35)
meshcentral  |     at endReadableNT (node:internal/streams/readable:1368:12)
meshcentral  |     at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
```

**Your config.json file**
```
{
  "$schema": "https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json",
  "__comment1__": "This is a simple configuration file, all values and sections that start with underscore (_) are ignored. Edit a section and remove the _ in front of the name. Refer to the user's guide for details.",
  "__comment2__": "See node_modules/meshcentral/sample-config-advanced.json for a more advanced example.",
  "settings": {
    "_maintenanceMode": true,
    "cert": "meshcentral.mydomain.com",
    "_WANonly": true,
    "_LANonly": true,
    "sessionKey": "blahblahblahblahblahblahblahblahblah",
    "sessionTime": 60,
    "port": 443,
    "_aliasPort": 443,
    "_redirPort": 80,
    "_redirAliasPort": 80,
    "_MongoDb": "mongodb://127.0.0.1:27017/meshcentral",
    "_AllowLoginToken": true,
    "_AllowFraming": true,
    "WebRTC": true,
    "_ClickOnce": false,
    "_UserAllowedIP" : "127.0.0.1,192.168.0.0/16"
  },
  "domains": {
    "": {
      "title": "Meshcentral",
      "_title2": "",
      "siteStyle": 2,
      "nightMode": 1,
      "_minify": true,
      "AutoRemoveInactiveDevices": 14,
      "DeviceSearchBarServerAndClientName": true,
      "userSessionIdleTimeout": 60,
      "newAccounts": false,
      "_userNameIsEmail": true,
      "authStrategies": {
        "__comment__": "This section is used to allow users to login using other accounts. You will need to get an API key from the services and register callback URL's",
        "azure": {
          "newAccounts": true,
          "clientid": "blahblahblahblahblahblahblahblahblah",
          "clientsecret": "blahblahblahblahblahblahblahblahblah",
          "tenantid": "blahblahblahblahblahblahblahblahblah"
        }
      }
    }
  },
  "letsencrypt": {
    "__comment__": "Requires NodeJS 8.x or better, Go to https://letsdebug.net/ first before trying Let's Encrypt.",
    "email": "admin@mydomain.com",
    "names": "meshcentral.mydomain.com",
    "skipChallengeVerification": false,
    "production": true
  }
}
```

**Comment by buckybytes** [comment]
Sorry for the long delay.  It looks like Azure isn't a common use case around here.

This is almost certainly an Azure config issue, as MeshCentral isn't handling any authentication at this stage.

If you're getting this error, Azure is accepting the username and password, but failing because of one of the following rules in the Azure configuration, listed in the order they are processed:

Identity protection policies
MFA
Approved client app / app protection policy
Trusted device
Custom controls
Session controls

Let us know if you figured it out, and what it was so we can help the next user in a more timely fashion.

**Comment by naaador** [comment]
I just came across this issue myself over the weekend while I was traveling. I normally access MeshCentral from my office where there's an exception for our static IPs in our CAP so assumed everything was working as expected. When outside the office MFA was required and I was seeing the same behavior where I'd get redirected directly to MeshCentral after entering my password, followed by an Internal Server Error and the same Exception in the logs as mentioned above, the MFA prompt was never displayed. For reference we're using Azure/Entra MFA.

After much trial and error I determined that if I logged into any other application first (ie directly into the Microsoft 365 portal, or any of our other OIDC or SAML applications), then logged into MeshCentral I could log in successfully. I compared the redirection URL used by another of our OIDC applications to what MeshCentral was using (when configured with the "azure" authStrategy) and noticed the URL was different. MeshCentral was using https://login.microsoft.net/<tenantid>/oauth2/authorize (going by memory on this one), while my other applications used https://login.microsoftonline.com/<tenantid/oauth2/v2/authorize.

More research lead me to [https://ylianst.github.io/MeshCentral/meshcentral/openidConnectStrategy/#azure-preset](https://ylianst.github.io/MeshCentral/meshcentral/openidConnectStrategy/#azure-preset) which uses the "oidc" authStrategy and describes the latter url from above. I switched to this authStrategy and testing verified that this works as expected.

TLDR
Seems the "azure" authStrategy doesn't work properly with MFA, use the "oidc" authStrategy with the "azure" preset. See [https://ylianst.github.io/MeshCentral/meshcentral/openidConnectStrategy/#azure-preset](https://ylianst.github.io/MeshCentral/meshcentral/openidConnectStrategy/#azure-preset) for details on how to configure.
